desc/protoparse/linker.go

```diff
@@ -1152,32 +1154,17 @@ func (l *linker) checkJsonNamesInMessage(md *desc.MessageDescriptor, res *parseR
 }
 
 func (l *linker) checkJsonNamesInEnum(ed *desc.EnumDescriptor, res *parseResult) error {
-	type jsonName struct {
-		source *dpb.EnumValueDescriptorProto
-		// enum's camel-case name (which can differ in case from map key)
-		orig string
-	}
-	seen := map[string]jsonName{}
+	seen := map[string]*dpb.EnumValueDescriptorProto{}
 	for _, evd := range ed.GetValues() {
 		scope := "enum value " + ed.GetName() + "." + evd.GetName()
 
-		ccName := internal.JsonName(evd.GetName())
-		lcaseEnumName := strings.ToLower(ed.GetName())
-		lcaseName := strings.TrimPrefix(strings.ToLower(ccName), lcaseEnumName)
-		if len(lcaseName) != len(ccName) {
-			// it had a prefix of enum name; also strip from mixed-case string
-			ccName = ccName[len(lcaseEnumName):]
-		}
-		if existing, ok := seen[lcaseName]; ok && evd.GetNumber() != existing.source.GetNumber() {
+		name := canonicalEnumValueName(evd.GetName(), ed.GetName())
```

Comment:
After playing with a few extra edge cases with `protoc`, I realized that the algorithm was not _quite_ what I had formulated here...

So I had to go look at the protoc C++ source code to reproduce it properly.